games/datingsim/master.py

```diff
@@ -109,101 +159,163 @@ def setup(self, **game_instance) -> None:
+                answer_b = self.get_answer(self.player_b)
+
+                # check if player b gives correct response
+                is_valid_turn = self.check_validity(answer_b)
+                self.proceed = is_valid_turn
+                self.last_message = self.update_answer(answer_b)
+
             self.current_turn += 1
+            self.complete_turns += 1
 
-            # Step 3,5,7...n
-            # GM to P2
-            # Gives message from P1 to P2
-            self.add_message(self.player_b, utterance=self.player_a.history[-1])
+            if self.current_turn > self.n_turns:
+                action = {'type': 'metadata', 'content': 'Game unsuccessful. Out of turns'}
+                self.log_event(from_='GM', to='GM', action=action)
+                self.proceed = False
 
-            # P2 to GM
-            # Answers to P1's message
-            self.get_answer(self.player_b)
-            self.log_next_turn()
-            self.current_turn += 1
 
-    print("end")
+    def check_validity(self, answer):
+        """
+        Check if given answer by yplayer is valid or
+        if it must be re-entered.
+        """
+        # check, if answer begins and ends with 
+        # pattern_for_answer = r"\[reason\].+\[end\]\n\[sentiment\].+\[end\]\n\[response\].+\[end\]"
+        # pattern_for_answer = r".+\[reason\].+"
+        # import re
+        # if re.fullmatch(pattern_for_answer, answer, re.DOTALL):
+        #     pass
+
+        # else: # abort game 
+        #     self.aborted = True
+            
+        #     # log the abortion event
+        #     action = {'type': 'invalid format', 'content': 'Aborted'}
+        #     self.log_event(from_='GM', to='GM', action=action)
+        #     logger.info(f"Invalid format.")
+
+        #     # increase the counter of requests that violate form rules
+        #     self.violated_request_counts[self.current_turn] += 1
+        #     return False
+        
+        # increase the counter of requests that conform to form rules
+        self.parsed_request_counts[self.current_turn] += 1
+        # log the event that the string was valid (no strange characters)
+        action = {'type': 'metadata', 'content': 'valid string'}
+        self.log_event(from_='GM', to='GM', action=action)
+
+        # log the fact that the answer was correct
+        action = {'type': 'parse',
+                  'content': f'{answer} conforms to rules'}
+
+        self.log_event(from_='GM', to='GM', action=action)
+        return True
+
+    def update_answer(self, answer):
+        """
+        Update the last response said
+        """
+        # filter out the response 
+        response_pattern = r"\[response\](.+)"
+        response_match = re.search(response_pattern, answer, re.DOTALL)
+        last_message = response_match.group(1)
+        # print(f"Last message: {last_message}")
+        return last_message
+
 
+    
 
 # This needs to be adjusted or removed completely (replaced)
```

Comment:
Can we delete this function? or should we refactor it?

Reply:
I would like to keep them commented in there for now until we have a working re-prompting and enforcing the template. As a guidence or reference while working on it. After that, we can gladly delete them.